### ISSUE TITLE
Zooming into fully visible canvas map causes iPhone 7's Safari to crash

### ISSUE THREAD

**Issue by jbazalar** [Leaflet/Leaflet#5496]
## How to reproduce
- Leaflet version I'm using: 1.0.3
- Browser (with version) I'm using: Safari or Chrome
- OS/Platform (with version) I'm using: iOS 10.3.1, issue only on iPhone 7 or iPhone 7 Plus 
- Zooming into fully visible canvas map causes iPhone 7's Safari to crash

## What behaviour I'm expecting and which behaviour I'm seeing
Expecting:
- I should be able to pinch zoom in to a map that fully covers the viewport and not have any issue
Results:
- When pinch zooming in (or out) while the map is covering the entire viewport, Safari (and Chrome) will crash, but only when this is running on iPhone 7 or 7 Plus.

## Minimal example reproducing the issue
The following occurs on a custom map, but is also seen on the leafletjs.com example maps.
- Open Safari on iPhone 7 (Plus), not on simulator (or browserstack).
- Go to http://leafletjs.com
- Scroll to area below the example map. Map should remain be visible.
- Zoom-in as much as the page can zoom using the empty/white space below the map and above "Here we create..."
- Once you can't zoom in anymore, scroll the page back up using the empty space so only the map canvas is in the viewport.
- At this point you'll pinch zoom in or out (you may need to zoom in or our a few levels).
- Then Safari should crashes (and reloads. Chrome will simply show the 'Aww Snap' page.

## NOTES:
- I've tried to replicate this on other versions of the iPhone (6) with the same iOS version, but cannot replicate the crash.
- Issue does not happen with Android.
- I've tried to get console information the moment the map crashes while the iPhone 7 is connected to a Mac (for inspector), but the inspector window closes when it crashes and reloads the page.
- First encountered using the leaflet.fullscreen plugin (mapbox), which zooms in the map to the viewport area, but was able to target it to the base map and replicated on a simple version of the map (thus the leafletjs.com example). Meaning it wasn't the plugin causing the issue.

**Comment by perliedman** [comment]
I don't know if any maintainer has access to an iPhone 7 Plus, so this is probably going to require someone else to step up and investigate what is going on.

**Comment by hypeitinc** [comment]
I recently started looking at Leaflet for a project so I'm a bit new here... 
As I have an iPhone 7 I tried to follow the instructions to replicate the problem on Safari, but I can't. No crashes. My phone is also on iOS 10.3.1.

**Comment by perliedman** [comment]
@hypeitinc hi and thanks for looking into this!

It's somewhat encouraging to hear that this might related to a specific device. On the other hand, if it turns out to occur in a significant percentage of iPhone 7 but not all, it will likely be a nightmare to debug :)

**Comment by jbazalar** [comment]
Okay, I recorded this earlier today as seen on an iPhone 7 Plus (https://www.youtube.com/watch?v=2uho37HYeqE) based on the replication steps noted. I have tried this same test on different iPhone 7 Plus and on a regular iPhone 7.  Sometimes, it takes more zoom ins or outs, sometimes it's it only needs to be really zoomed in prior to moving the map into the viewing area and it'll error out on it's own.

**Comment by jbazalar** [comment]
Did the example video help you replicate the issue? (https://www.youtube.com/watch?v=2uho37HYeqE)

**Comment by perliedman** [comment]
I found an iPhone 7 plus and could reproduce this, it look like it's when the popup comes into view something weird happens.

Still, this is _really_ an edge case, and personally this will end up very low on the looong list of issues in Leaflet. If this issue is important to you, help with debugging and fixing is very welcome.

**Comment by Ernie23** [comment]
Hi evereybody,

I can reproduce the same problem on different websites using Leaflet 1.0.3.
In my case I am using my IPAD Gen.3 with iOS v. 9.3.5.

I hope you can reproduce this with my fiddle:
- Open Safari on iPad
- Go to https://jsfiddle.net/zdp195ck/1/
- Zoom-in as much as the page can zoom using gray JavaScirpt block
- Once you can't zoom in anymore, scroll the page left until tha map appears in the viewport.
- At this point you'll pinch zoom in or out few zoom levels

Result: Website crashes and Safari reload the website.
            Following message came from Safari "The website was reloaded because there was a problem."

I will try to investigate further.

Markus

**Comment by zspitzer** [comment]
here is the relevant webkit bug https://bugs.webkit.org/show_bug.cgi?id=172206 
I have referenced this issue

**Comment by Ernie23** [comment]
@zspitzer This is not only on iOS10. I can reproduce it on Safari 9 / iOS 9.3.5 with my iPad Gen.3.
Can you edit your bug report https://bugs.webkit.org/show_bug.cgi?id=172206 ?

**Comment by zspitzer** [comment]
Why does that matter, 9.3.5 is no longer supported and already riddled with major security holes?

**Comment by Ernie23** [comment]
@zspitzer Many of other customers using iPad Gen.3 devices. I not 100% sure, but I think iPad 3 is not compatible with iOS10. Am i right?

**Comment by zspitzer** [comment]
Yep, only 4th gen iPads have limited iOS 10 support https://en.wikipedia.org/wiki/IOS_version_history#iOS_10

**Comment by zspitzer** [comment]
I'm not familiar with the Leaflet code myself, does leaflet create a lot of canvas objects? perhaps reusing a single canvas object might help, or adding a setTimeout(fn, 0) approach might allow the GC to kick in?

**Comment by Ernie23** [comment]
@zspitzer no in our solutions, there is no canvas in use. In my fiddle above leaflet only load tiles from a tileserver. jbazalar shows same error using first simple leaflet tutorial.

**Comment by perliedman** [comment]
@zspitzer great to hear! A crashing browser does indeed indicate a bug in the browser and not in Leaflet itself.

As @Ernie23 already said, Leaflet normally does not use canvas elements, although it can under some circumstances. The default page most certainly does not use canvas.

**Comment by zspitzer** [comment]
Maybe once Apple fixes this in an upcoming release of Safari they might be able to advise workarounds for older versions? 

Also, you can find crash reports on iOS by following this guide (tho it's under Privacy on iOS 10)
https://www.chromium.org/developers/how-tos/retrieving-crash-reports-on-ios

You might see something like reason:vm-thrashing as a cause

**Comment by fetis** [comment]
We have the same problem with our SDK based on Leaflet. After pinch-in, pause, pinch-out Safari crashes or reloads the page. I've checked the Developer Tools and a lot of usage of Layers memory, this's because Leaflet uses 3d transformations. I don't know is it possible to reduce this memory load.

**Comment by perliedman** [comment]
@fetis if you can reproduce it on apple.com, it does not sound like a Leaflet issue, right?

FWIW, you can disable 3D transforms by setting [`L_DISABLE_3D`](http://leafletjs.com/reference-1.2.0.html#global-switches) before initalizing the map.

**Comment by fetis** [comment]
@perliedman it's definitely not the Leaflet issue, but Leaflet is affected by it because it uses 3D transformations a lot. I was able to reproduce this in a code snippet and filled a detailed report here https://bugs.webkit.org/show_bug.cgi?id=172206#c4

From Leaflet side, we can reduce pressure on Layers memory(GPU memory for compositing and painting) but what actually needs to be done for this I'm not sure.

**Comment by chris-aeviator** [comment]
I can confirm that the solution proposed by @perliedman is solving this issue for all iOS devices that crashed on our app using ember-leaflet (leaflet.js) even though it feels more sluggy to use :+1:

**Comment by SandySJadhav** [comment]
Hi there. 

I have similar king of issue while using canvas. My app was crashing whenever I tried to use canvas inside iframes. Basically canvas add extra layer on itself whenever we draw/clear canvas. This causes memory leackage in safari browser for iOS devices. When I remove this canvas element, no any **Pinch zoom crash happn**. 
I am still looking for solution/alternative for canvas. Hope this will help you find clues in your project
